### PR TITLE
PYIC-7329: Refactor core-front tests

### PR DESCRIPTION
**Pull request by MikeCollingwood**
### What changed

- Standardise core-front tests
- Use mock-express test util to commonise how we defined req & res in core-front tests
- Use `sinon.fake`'s wherever possible as is the simplest class of mock
- Inline variables wherever sensible, to be consistent and it is more concise.
- Added in Arrange, Act and Assert comments to help readability

### Why did it change

- Improve readability & maintainability

### Issue tracking

- [PYIC-7329](https://govukverify.atlassian.net/browse/PYIC-7329)

[PYIC-7329]: https://govukverify.atlassian.net/browse/PYIC-7329?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/app/ipv/tests/checkFormRadioButtons.test.ts [hunk 1]
```diff
@@ -1,39 +1,37 @@
...
+  specifyCreateResponse,
+} from "../../../test-utils/mock-express";
 
 describe("checkFormRadioButtonSelected middleware", () => {
-  const testReq = {
-    body: {},
+  // Mock handler parameters
+  const createRequest = specifyCreateRequest({
     params: { pageId: "page-ipv-identity-document-start" },
-    csrfToken: sinon.fake(),
```
Comment (MikeCollingwood) [inline src/app/ipv/tests/checkFormRadioButtons.test.ts:9]: Does not need these defined as default in specifyCreateRequest

File: src/handlers/error-handler.test.ts [hunk 2]
```diff
@@ -1,171 +1,240 @@
...
-      render: sinon.fake(),
-    } as any;
-
-    next = sinon.fake() as any;
-  });
-
-  afterEach(() => {
-    sinon.restore();
+    next.resetHistory();
+    axiosStub.isAxiosError.reset();
```
Comment (MikeCollingwood) [inline src/handlers/error-handler.test.ts:31]: dubious

File: src/app/credential-issuer/middleware.ts [hunk 3]
```diff
@@ -44,7 +44,7 @@ export const sendParamsToAPI: RequestHandler = async (req, res) => {
   }
 
   const apiResponse = await postCriCallback(req, body);
-  return handleBackendResponse(req, res, apiResponse?.data);
+  return handleBackendResponse(req, res, apiResponse.data);
```
Comment (MikeCollingwood) [inline src/app/credential-issuer/middleware.ts:47]: `data` is not an optional parameter on AxiosResponse, the return type of postCriCallback

File: src/test-utils/mock-express.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  session?: Record<string, unknown>;
+  method?: string;
+  path?: string;
+  id?: string;
+  ip?: string;
+  cookies?: Record<string, unknown>;
+}
+
+export const specifyCreateRequest =
+  (commonParams: CreateRequestParams = {}) =>
```
Comment (DanCorderIPV) [inline src/test-utils/mock-express.ts]: I think calling this `defaultParams` might make it a little easier to understand

File: src/test-utils/mock-express.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+  (params: CreateRequestParams = {}): Request => {
```
Comment (DanCorderIPV) [inline src/test-utils/mock-express.ts:19]: This is the linter's fault, but I really don't like how it's arranged this code :(

File: src/services/parameterStoreService.test.ts [hunk 6]
```diff
@@ -56,18 +64,20 @@ describe("parameterStoreService", () => {
...
+      expect(param).to.deep.include({
         pageId: "/some-page",
         bannerMessage: "Test banner",
         bannerMessageCy: "Welsh Test banner",
       });
       expect(
         ssmClientStub.calledWithMatch({
           Name: "/core-front/test-parameter",
-        }),
+        } as any),
```
Comment (DanCorderIPV) [inline src/services/parameterStoreService.test.ts]: This doesn't seem to be needed

File: src/app/ipv/tests/handleJourneyActionRequest.test.ts [hunk 7]
```diff
@@ -2,161 +2,178 @@ import { expect } from "chai";
...
+    expect(
+      coreBackServiceStub.postJourneyEvent,
+    ).to.have.been.calledOnceWithExactly(
+      req,
+      "some-journey-event",
+      req.session.currentPage,
+    );
   });
 
   it("should postJourneyEvent and use ip address from header when not present in session", async function () {
```
Comment (DanCorderIPV) [inline src/app/ipv/tests/handleJourneyActionRequest.test.ts]: This test and the one below are supposed to check that the correct IP address is being chosen - so both IP addresses should still be specified in the tests.
However, I can't see that the IP address is checked anywhere in the test, so perhaps these tests are obsolete?

Comment (DanCorderIPV) [reply src/app/ipv/tests/handleJourneyActionRequest.test.ts]: Looks like Thea added the original test on 21/10/24. Could you please talk to her and see if these tests are doing what they are supposed to?

File: src/app/ipv/tests/handleJourneyActionRequest.test.ts [hunk 8]
```diff
@@ -2,161 +2,178 @@ import { expect } from "chai";
...
+    ).to.have.been.calledOnceWithExactly(
+      req,
+      req.body.journey,
+      req.session.currentPage,
+    );
   });
 
-  it("should call redirect given 'contact' event", async function () {
-    const req = { ...testReq, body: { journey: "contact" } };
+  it("should use session IP if present", async function () {
```
Comment (DanCorderIPV) [inline src/app/ipv/tests/handleJourneyActionRequest.test.ts]: Again there doesn't appear to be any testing of the IP address - so what is this test for?

**Review by DanCorderIPV — DISMISSED** [review]

File: src/app/ipv/tests/handleJourneyActionRequest.test.ts (hunk 7 above)
Comment (MikeCollingwood) [reply src/app/ipv/tests/handleJourneyActionRequest.test.ts]: It looks like this was the last thought on the matter: https://github.com/govuk-one-login/ipv-core-front/pull/1611#discussion_r1809049591. This (and the beneath) test looks redundant as the logic occurs within the postJourneyEvent method, which has a test covering it:
`coreBackService.test.ts > "should postJourneyEvent with correct parameters and headers"`.

**Comment by DanCorderIPV** [comment]
Looks good, I can give it a tick once the merge conflicts are fixed.

**Review by DanCorderIPV — DISMISSED** [review]

**Review by DanCorderIPV — APPROVED** [review]

**Pull request merged**